File: src/main/java/jp/kentan/minecraft/neko_event/module/key/KeyManager.java
package jp.kentan.minecraft.neko_event.module.key;

import jp.kentan.minecraft.neko_event.NekoEvent;
import jp.kentan.minecraft.neko_event.config.listener.ConfigListener;
import jp.kentan.minecraft.neko_event.config.provider.KeyConfigProvider;
import jp.kentan.minecraft.neko_event.module.key.model.Key;
import jp.kentan.minecraft.neko_event.util.Log;
import jp.kentan.minecraft.neko_event.util.NekoUtil;
import org.bukkit.ChatColor;
import org.bukkit.Location;
import org.bukkit.Material;
import org.bukkit.command.CommandSender;
import org.bukkit.entity.Player;
import org.bukkit.inventory.ItemStack;
import org.bukkit.plugin.java.JavaPlugin;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class KeyManager implements ConfigListener<Key>{
    private static KeyManager sInstance = new KeyManager();

    private static Map<String, Key> sKeyMap = new HashMap<>();

    public static void setup(JavaPlugin plugin){
        KeyConfigProvider.bindListener(sInstance);
        KeyConfigProvider.load();

        plugin.getCommand("key").setExecutor(new KeyCommandExecutor());
    }

    static void create(Player player, String id, Map<String, List<String>> optionMap){
        final ItemStack itemStack = player.getInventory().getItemInMainHand();

        if(itemStack == null || itemStack.getType() == Material.AIR){
            player.sendMessage(NekoEvent.PREFIX + ChatColor.YELLOW + "登録するアイテムを持って下さい.");
            return;
        }

        final Option option = Option.analyze(player, optionMap);

        if(option == null){
            return;
        }

        if(sKeyMap.containsKey(id)){
            player.sendMessage(KEY_ALREADY_REGISTERED_TEXT.replace("{id}", id));
            return;
        }

        final Key key = KeyConfigProvider.create(id, new ItemStack(itemStack), (option.IS_TAKE != null) ? option.IS_TAKE : false,
                option.PERIOD, option.BLOCK_MATERIAL, option.BLOCK_LOCATION,
                option.MATCH_MSG, option.NOT_MATCH_MSG, option.EXPIRED_MSG, option.SHORT_AMOUNT_MSG);

        if(key == null){
            player.sendMessage(NekoEvent.PREFIX + ChatColor.YELLOW + "作成に失敗しました.");
        }else{
            Log.info("{player}がキー({id})を{item}で作成しました.".replace("{player}", player.getName()).replace("{id}",id).replace("{name}", key.getName()));
            sKeyMap.put(id, key);
        }
    }

    static void delete(CommandSender sender, String id){
        if(!sKeyMap.containsKey(id)){
            sender.sendMessage(KEY_NOT_FOUND_TEXT.replace("{id}", id));
            return;
        }

        if(KeyConfigProvider.delete(id)){
            sKeyMap.remove(id);
            Log.warn("{sender}がキー({id})を消去しました.".replace("{sender}", sender.getName()).replace("{id}",id));
        }else{
            sender.sendMessage(NekoEvent.PREFIX + ChatColor.YELLOW + "消去に失敗しました.");
        }
    }

    static void setOptions(Player player, String id, Map<String, List<String>> optionMap){
        final Key key = sKeyMap.get(id);

        if(key == null){
            Log.error(KEY_NOT_FOUND_TEXT.replace("{id}", id));
            return;
        }

        final Option option = Option.analyze(player, optionMap);

        if(option == null){
            return;
        }

        if(key.update(null, option.IS_TAKE, option.PERIOD, option.BLOCK_MATERIAL, option.BLOCK_LOCATION,
                option.MATCH_MSG, option.NOT_MATCH_MSG, option.EXPIRED_MSG, option.SHORT_AMOUNT_MSG)){
            Log.info(player.getName() + "がキー({id})を更新しました.".replace("{id}", id));
        }else{
            player.sendMessage(NekoEvent.PREFIX + ChatColor.YELLOW + "キー({id})の更新に失敗しました.".replace("{id}", id));
        }
    }

    static void overwriteKeyItem(Player player, String id){
        final Key key = sKeyMap.get(id);

        if(key == null){
            Log.error(KEY_NOT_FOUND_TEXT.replace("{id}", id));
            return;
        }

        final ItemStack itemStack = player.getInventory().getItemInMainHand();

        if(itemStack == null || itemStack.getType() == Material.AIR){
            player.sendMessage(NekoEvent.PREFIX + ChatColor.YELLOW + "登録するアイテムを持って下さい.");
            return;
        }

        if(key.update(new ItemStack(itemStack), null, 0, null, null,
                null, null, null, null)){
            Log.info(player.getName() + "がキー({id})のアイテムを{name}で上書きしました.".replace("{id}", id).replace("{name}", key.getName()));
        }else{
            player.sendMessage(NekoEvent.PREFIX + ChatColor.YELLOW + "キー({id})の更新に失敗しました.".replace("{id}", id));
        }
    }

    private static void give(Player player, String id, int amount){
        if(!sKeyMap.containsKey(id)){
            Log.error("キー({id})は存在しません.".replace("{id}",id));
            return;
        }

        player.getInventory().addItem(sKeyMap.get(id).getItemStack(amount));
    }

    static void give(String playerName, String id, String strAmount){
        Player player = NekoUtil.toPlayer(playerName);

        if(player == null){
            return;
        }

        int amount = 1;

        if(strAmount != null) {
            amount = NekoUtil.toInteger(strAmount);

            if (amount < 1) {
                player.sendMessage(NekoEvent.PREFIX + ChatColor.YELLOW + "個数は1以上に設定して下さい.");
                return;
            }
        }

        give(player, id, amount);
    }

    private static void drop(Location location, String id, int amount){
        if(!sKeyMap.containsKey(id)){
            Log.error("キー({id})は存在しません.".replace("{id}",id));
            return;
        }

        location.getWorld().dropItem(location, sKeyMap.get(id).getItemStack(amount));
    }

    static void drop(String id, String worldName, String strX, String strY, String strZ, String strAmount){
        Location location = NekoUtil.toLocation(worldName, strX, strY, strZ);

        if(location == null){
            return;
        }

        int amount = 1;

        if(strAmount != null) {
            amount = NekoUtil.toInteger(strAmount);

            if (amount < 1) {
                Log.error(NekoEvent.PREFIX + ChatColor.YELLOW + "個数は1以上に設定して下さい.");
                return;
            }
        }

        drop(location, id, amount);
    }


    private static void use(Player player, String id){
        Key key = sKeyMap.get(id);

        if(key == null){
            Log.error("キー({id})は存在しません.".replace("{id}",id));
            return;
        }

        ItemStack handItem = player.getInventory().getItemInMainHand();

        switch (key.match(handItem)){
            case MATCH:
                if(key.isTake()){
                    int amount = handItem.getAmount() - key.getAmount();
                    player.getInventory().getItemInMainHand().setAmount(amount);
                }

                if(key.hasBlockOption()){
                    key.getBlockLocation().getBlock().setType(key.getBlockMaterial());
                }

                if(key.hasMatchMessage()){
                    player.sendMessage(key.getMatchMessage());
                }
                break;
            case NOT_MATCH:
                if(key.hasNotMatchMessage()){
                    player.sendMessage(key.getNotMatchMessage());
                }
                break;
            case EXPIRED:
                if(key.hasExpiredMessage()){
                    player.sendMessage(key.getExpiredMessage());
                }
                break;
            case SHORT_AMOUNT:
                if(key.hasShortAmountMessage()){
                    player.sendMessage(key.getShortAmountMessage());
                }
                break;
            default:
        }
    }

    static void use(String playerName, String id){
        Player player = NekoUtil.toPlayer(playerName);

        if(player == null){
            return;
        }

        use(player, id);
    }

    static void sendList(CommandSender sender){
        StringBuilder builder = new StringBuilder(ID_LIST_TEXT + "\n");

        sKeyMap.entrySet().stream()
                .sorted(Map.Entry.comparingByKey())
                .forEach(e -> {
                            builder.append(" ");
                            builder.append(e.getKey());
                            builder.append(": ");
                            builder.append(e.getValue().getName());
                            builder.append('\n');
                            builder.append(ChatColor.RESET);
                        }
                );

        sender.sendMessage(builder.toString());
    }

    static void sendInfo(CommandSender sender, String id){
        final Key key = sKeyMap.get(id);

        if(key == null){
            sender.sendMessage(KEY_NOT_FOUND_TEXT.replace("{id}",id));
            return;
        }

        sender.sendMessage(INFO_TEXT);
        sender.sendMessage(" ID: " + id);
        sender.sendMessage(" 名前: " + key.getName());
        if(key.hasBlockOption()){
            sender.sendMessage(" 設置ﾌﾞﾛｯｸ: " + key.getBlockMaterial());
            sender.sendMessage(" 設置位置: " + NekoUtil.toString(key.getBlockLocation()));
        }
    }

    static void sendOptionHelp(CommandSender sender){
        sender.sendMessage("---------- NekoEvent Keyオプションヘルプ ----------");
        sender.sendMessage("| " + ChatColor.AQUA + "TAKE [true/false]               (解除時にキーを消す)");
        sender.sendMessage("| " + ChatColor.AQUA + "PERIOD [minutes]                (キーの有効期限)");
        sender.sendMessage("| " + ChatColor.AQUA + "BLOCK_MATERIAL [material name]  (解除時に設置するブロック)");
        sender.sendMessage("| " + ChatColor.AQUA + "BLOCK_LOCATION [HERE/x y z]     (解除時に設置する場所)");
        sender.sendMessage("| " + ChatColor.AQUA + "MATCH_MSG [message]           (一致時に表示)");
        sender.sendMessage("| " + ChatColor.AQUA + "NOT_MATCH_MSG [message]       (不一致時に表示)");
        sender.sendMessage("| " + ChatColor.AQUA + "EXPIRED_MSG [message]           (期限切れ時に表示)");
        sender.sendMessage("| " + ChatColor.AQUA + "SHORT_AMOUNT_MSG [message]      (数が足りない時に表示)");
        sender.sendMessage("| " + ChatColor.GRAY + "[]は必須,<>は任意,()は説明です.");
        sender.sendMessage("| " + ChatColor.GRAY + "[message]では{name}でキー名に置換されます.");
        sender.sendMessage("---------------------------------------");
    }

    @Override
    public void onUpdate(Map<String, Key> dataMap) {
        sKeyMap.clear();
        sKeyMap.putAll(dataMap);
    }

    private final static String KEY_NOT_FOUND_TEXT = NekoEvent.PREFIX + ChatColor.YELLOW + "キー({id})は存在しません.";
    private final static String KEY_ALREADY_REGISTERED_TEXT = NekoEvent.PREFIX + ChatColor.YELLOW + "このID({id})はすでに登録されています.";

    private final static String ID_LIST_TEXT = ChatColor.translateAlternateColorCodes('&', "&7--------- &6キー一覧 &7---------&r");
    private final static String INFO_TEXT = ChatColor.translateAlternateColorCodes('&', "&7--------- &6キー情報 &7---------");


    static class Option {
        final static String TAKE_KEY = "TAKE";
        final static String PERIOD_KEY = "PERIOD";
        final static String BLOCK_MATERIAL_KEY = "BLOCK_MATERIAL";
        final static String BLOCK_LOCATION_KEY = "BLOCK_LOCATION";
        final static String MATCH_MSG_KEY = "MATCH_MSG";
        final static String NOT_MATCH_MSG_KEY = "NOT_MATCH_MSG";
        final static String EXPIRED_MSG_KEY = "EXPIRED_MSG";
        final static String SHORT_AMOUNT_MSG_KEY = "SHORT_AMOUNT_MSG";

        private Boolean IS_TAKE;
        private int PERIOD = 0;
        private Material BLOCK_MATERIAL;
        private Location BLOCK_LOCATION;
        private String MATCH_MSG, NOT_MATCH_MSG, EXPIRED_MSG, SHORT_AMOUNT_MSG;


        static Option analyze(Player player, Map<String, List<String>> optionMap) {
            Option option = null;

            try {
                option = new Option(player, optionMap);
            } catch (Exception e) {
                e.printStackTrace();
                Log.error(e.getMessage());
            }

            return option;
        }

        private Option(Player player, Map<String, List<String>> map) throws Exception {
            MATCH_MSG      = NekoUtil.appendStrings(map.remove(MATCH_MSG_KEY));
            NOT_MATCH_MSG  = NekoUtil.appendStrings(map.remove(NOT_MATCH_MSG_KEY));
            EXPIRED_MSG      = NekoUtil.appendStrings(map.remove(EXPIRED_MSG_KEY));
            SHORT_AMOUNT_MSG = NekoUtil.appendStrings(map.remove(SHORT_AMOUNT_MSG_KEY));


            if(map.containsKey(TAKE_KEY)){
                IS_TAKE = map.remove(TAKE_KEY).get(0).equals("true");
            }

            if(map.containsKey(PERIOD_KEY)){
                PERIOD = NekoUtil.toInteger(map.remove(PERIOD_KEY).get(0));
            }

            List<String> location = new ArrayList<>();
            if(map.containsKey(BLOCK_LOCATION_KEY)) {
                location.addAll(map.remove(BLOCK_LOCATION_KEY));

                if (location.size() > 0) {
                    if (location.get(0).equals("HERE")) {
                        BLOCK_LOCATION = player.getLocation();
                    } else if (NekoUtil.checkParams(location.size(), 3, BLOCK_LOCATION_KEY)) {
                        BLOCK_LOCATION = NekoUtil.toLocation(
                                player.getLocation(),
                                new String[]{
                                        location.get(0),
                                        location.get(1),
                                        location.get(2)
                                }
                        );
                    }
                }
            }

            if(map.containsKey(BLOCK_MATERIAL_KEY)){
                String materialName = map.remove(BLOCK_MATERIAL_KEY).get(0);
                BLOCK_MATERIAL = Material.matchMaterial(materialName);

                if(BLOCK_MATERIAL == null){
                    Log.error("マテリアル(" + materialName + ")は存在しません.");
                    throw new Exception("Invalid material name.");
                }

                if(!BLOCK_MATERIAL.isBlock()){
                    Log.error("マテリアル(" + materialName + ")はブロックではありません.");
                    throw new Exception("Invalid material name.");
                }
            }

            if (map.size() > 0) {
                map.forEach((key, val) -> Log.error("オプション({key})は存在しません.".replace("{key}", key)));
                throw new Exception("Invalid option.");
            }
        }
    }
}

File: src/main/java/jp/kentan/minecraft/nekoevent/manager/KeyManager.kt
package jp.kentan.minecraft.nekoevent.manager

import jp.kentan.minecraft.nekoevent.NekoEvent
import jp.kentan.minecraft.nekoevent.component.KeyFlag
import jp.kentan.minecraft.nekoevent.component.KeyFlag.*
import jp.kentan.minecraft.nekoevent.component.KeyResult
import jp.kentan.minecraft.nekoevent.component.model.Key
import jp.kentan.minecraft.nekoevent.config.ConfigUpdateListener
import jp.kentan.minecraft.nekoevent.config.provider.KeyConfigProvider
import jp.kentan.minecraft.nekoevent.util.*
import org.bukkit.ChatColor
import org.bukkit.Location
import org.bukkit.Material
import org.bukkit.command.CommandSender
import org.bukkit.entity.Player

class KeyManager(
        private val keyConfig: KeyConfigProvider
) : ConfigUpdateListener<Key> {

    private val keyMap = mutableMapOf<String, Key>()

    init {
        keyConfig.listener = this
    }

    fun getKey(keyId: String): Key? = keyMap[keyId]

    fun getKeyIdList() = keyMap.values.map { it.id }.sorted()

    fun give(strPlayer: String, keyId: String, strAmount: String) {
        val player = strPlayer.toPlayerOrError() ?: return
        val amount = strAmount.toIntOrError() ?: return

        give(player, keyId, amount)
    }

    fun give(player: Player, keyId: String, amount: Int) {
        val key = keyMap.getOrError(keyId) ?: return

        give(player, key, amount)
    }

    private fun give(player: Player, key: Key, amount: Int) {
        player.inventory.addItem(key.getItemStack(amount))
    }

    fun drop(keyId: String, strLocation: Array<String>, strAmount: String) {
        val key = keyMap.getOrError(keyId) ?: return

        val location = strLocation.toLocationOrError() ?: return
        val amount = strAmount.toIntOrError() ?: return

        drop(key, location, amount)
    }

    private fun drop(key: Key, location: Location, amount: Int) {
        location.world.dropItemNaturally(location, key.getItemStack(amount))
    }

    fun use(strPlayer: String, keyId: String) {
        val player = strPlayer.toPlayerOrError() ?: return

        use(player, keyId)
    }

    fun use(player: Player, keyId: String): Boolean {
        val key = keyMap.getOrError(keyId) ?: return false

        return use(player, key)
    }

    private fun use(player: Player, key: Key): Boolean {
        val handItem = player.inventory.itemInMainHand

        when (key.compareTo(handItem)) {
            KeyResult.MATCH -> {
                if (key.enabledTake) {
                    val amount = handItem.amount - key.amount
                    player.inventory.itemInMainHand.amount = amount
                }

                key.block?.let { (material, location) ->
                    location.block.type = material
                }

                key.formatMatchMessage?.let { player.sendMessage(it) }

                return true
            }
            KeyResult.NOT_MATCH -> key.formatNotMatchMessage?.let { player.sendMessage(it) }
            KeyResult.EXPIRED -> key.formatExpiredMessage?.let { player.sendMessage(it) }
            KeyResult.SHORT_AMOUNT -> key.formatShortAmountMessage?.let { player.sendMessage(it) }
        }

        return false
    }

    fun create(player: Player, keyId: String) {
        val itemStack = player.inventory.itemInMainHand
        if (itemStack == null || itemStack.type == Material.AIR) {
            player.sendMessage(NekoEvent.PREFIX + ChatColor.YELLOW + "登録するアイテムを持って下さい.")
            return
        }

        if (keyMap.containsKey(keyId)) {
            player.sendMessage(NekoEvent.PREFIX + ChatColor.YELLOW + "ID: ${keyId}は既に使用されています.")
            return
        }

        val key = Key(keyId, itemStack)
        if (keyConfig.update(key)) {
            keyMap[keyId] = key
            player.sendMessage(NekoEvent.PREFIX + key.name + ChatColor.GREEN + "を登録しました.")
        } else {
            player.sendMessage(NekoEvent.PREFIX + ChatColor.YELLOW + "保存に失敗しました.")
        }
    }

    fun delete(sender: CommandSender, keyId: String) {
        if (!keyMap.containsKey(keyId)) {
            sender.sendMessage(NekoEvent.PREFIX + ChatColor.YELLOW + "${keyId}は存在しません.")
            return
        }

        if (keyConfig.delete(keyId)) {
            keyMap.remove(keyId)
            sender.sendMessage(NekoEvent.PREFIX + ChatColor.GREEN + "${keyId}を消去しました.")
        } else {
            sender.sendMessage(NekoEvent.PREFIX + ChatColor.YELLOW + "消去に失敗しました.")
        }
    }

    fun flag(player: Player, keyId: String, flagId: String, flagArgs: List<String>) {
        var key = keyMap[keyId] ?: let {
            player.sendMessage(NekoEvent.PREFIX + ChatColor.YELLOW + "${keyId}は存在しません.")
            return
        }

        val flagType = KeyFlag.find(flagId) ?: let {
            player.sendMessage(NekoEvent.PREFIX + ChatColor.YELLOW + "${flagId}は正しいフラグIDではありません.")
            return
        }

        try {
            key = when (flagType) {
                ITEM                 -> key.copy(itemStack = flagArgs.flagItemStack(player))
                TAKE                 -> key.copy(enabledTake = flagArgs.flagBoolean(false))
                EXPIRE               -> key.copy(expireMinutes = flagArgs.flagInt(0))
                BLOCK_MATERIAL       -> key.copy(blockMaterial = flagArgs.flagMaterial(null))
                BLOCK_LOCATION       -> key.copy(blockLocation = flagArgs.flagLocation(player, null))
                MATCH_MESSAGE        -> key.copy(matchMessage = flagArgs.flagString(null))
                NOT_MATCH_MESSAGE    -> key.copy(notMatchMessage = flagArgs.flagString(null))
                EXPIRED_MESSAGE      -> key.copy(expiredMessage = flagArgs.flagString(null))
                SHORT_AMOUNT_MESSAGE -> key.copy(shortAmountMessage = flagArgs.flagString(null))
            }
        } catch (e: InvalidFlagException) {
            player.sendMessage(NekoEvent.PREFIX + ChatColor.YELLOW + e.message)
            return
        }

        if (keyConfig.update(key)) {
            keyMap[keyId] = key
            player.sendMessage(NekoEvent.PREFIX + ChatColor.GREEN + "$keyId を更新しました.")
        } else {
            player.sendMessage(NekoEvent.PREFIX + ChatColor.YELLOW + "更新に失敗しました.")
        }
    }

    fun sendList(sender: CommandSender) {
        val sb = StringBuilder("&7--------- &6鍵一覧 &7---------&r".formatColorCode())
        sb.append('\n')

        keyMap.toSortedMap().forEach { id, key ->
            sb.append(' ')
            sb.append(id)
            sb.append(": ")
            sb.append(key.name)
            sb.append('\n')
            sb.append(ChatColor.RESET)
        }

        sender.sendMessage(sb.toString())
    }

    fun sendInfo(sender: CommandSender, keyId: String) {
        val key = keyMap.getOrError(keyId) ?: return

        val messages = arrayOf(
                "&7--------- &6鍵情報 &7---------".formatColorCode(),
                " ID: ${key.id}",
                " 名前: ${key.name}",
                " 回収: ${key.enabledTake}",
                " 期限: ${if (key.enabledExpire) "${key.expireMinutes}分" else "なし" }",
                " ブロック: ${if (key.block != null) "${key.block.first}, ${key.block.second.formatString()}" else "なし"}",
                " 一致: ${key.formatMatchMessage}",
                " 不一致: ${key.formatNotMatchMessage}",
                " 期限切れ: ${key.formatExpiredMessage}",
                " 枚数不足: ${key.formatShortAmountMessage}"
        )

        sender.sendMessage(messages)
    }

    fun reload() {
        keyConfig.load()
    }

    override fun onConfigUpdate(dataMap: Map<String, Key>) {
        keyMap.clear()
        keyMap.putAll(dataMap)

        Log.info("${keyMap.size}個の鍵を読み込みました.")
    }

    private fun MutableMap<String, Key>.getOrError(id: String): Key? {
        return keyMap[id] ?: let {
            Log.error("鍵($id)は存在しません.")
            return@let null
        }
    }
}